File: android/app/src/main/kotlin/com/vervetogether/dreamvalley/DreamValleyAuthStorage.kt
package com.vervetogether.dreamvalley

import android.content.Context
import android.util.Log
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey

/**
 * Native auth-token storage backed by the Android Keystore.
 *
 * Mirror of iOS DreamValleyAuthStorage. Same MethodChannel contract:
 *   com.vervetogether.dreamvalley/auth  (Dart → Kotlin)
 *     "store" → { token: String } → Boolean (write-back-verified)
 *     "read"  → ()                → String? (or null on miss)
 *     "clear" → ()                → Boolean
 *
 * Storage: EncryptedSharedPreferences (androidx.security:security-crypto).
 *   - Master key in Android Keystore (hardware-backed where available),
 *     AES256_GCM key scheme.
 *   - Per-entry AES256_GCM for values + AES256_SIV for keys (deterministic
 *     key encryption is required so we can look up "session_token" by name).
 *   - Survives app updates. Uninstall wipes (by design — restore is the
 *     recovery path, NOT persistence across reinstall).
 *   - Min Android: 23 (M). Build script enforces this via Flutter minSdk.
 */
class DreamValleyAuthStorage(private val context: Context) {
    private val prefsName = "dv_auth_secure"
    private val tokenKey = "session_token"
    private val tag = "DVAuth"

    private fun prefs() = EncryptedSharedPreferences.create(
        context,
        prefsName,
        MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build(),
        EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
        EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
    )

    fun store(token: String): Boolean {
        if (token.isEmpty()) {
            Log.w(tag, "store: empty token rejected")
            return false
        }
        return try {
            val committed = prefs().edit().putString(tokenKey, token).commit()
            if (!committed) {
                Log.e(tag, "store: edit().commit() returned false")
                return false
            }
            // Write-back-verify: read it back and confirm match.
            val readBack = read()
            if (readBack == token) {
                Log.i(tag, "store: success (write-back verified, token_len=${token.length})")
                true
            } else {
                Log.e(tag, "store: write-back MISMATCH wrote_len=${token.length} read_len=${readBack?.length ?: -1}")
                false
            }
        } catch (e: Exception) {
            Log.e(tag, "store: exception ${e.message}", e)
            false
        }
    }

    fun read(): String? {
        return try {
            prefs().getString(tokenKey, null)
        } catch (e: Exception) {
            Log.e(tag, "read: exception ${e.message}", e)
            null
        }
    }

    fun clear(): Boolean {
        return try {
            val committed = prefs().edit().remove(tokenKey).commit()
            if (!committed) {
                Log.e(tag, "clear: edit().commit() returned false")
                false
            } else {
                Log.i(tag, "clear: success")
                true
            }
        } catch (e: Exception) {
            Log.e(tag, "clear: exception ${e.message}", e)
            false
        }
    }
}

File: ios/Runner/DreamValleyAuthStorage.swift
import Flutter
import Foundation
import Security

/// Native auth-token storage backed by the iOS Keychain.
///
/// MethodChannel: com.vervetogether.dreamvalley/auth  (Dart → iOS)
///   "store" → { token: String } → Bool (true iff write succeeded AND
///                                       immediate read-back returned an
///                                       identical token; write-back-verify)
///   "read"  → ()                 → String? (the stored token, nil on miss)
///   "clear" → ()                 → Bool   (true on cleared or already absent)
///
/// Accessibility: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
///   - Item readable only after the device has been unlocked at least once
///     since boot. Survives backgrounding.
///   - "ThisDeviceOnly" → never syncs to iCloud Keychain. Token is bound to
///     this physical install on this device.
///   - Uninstall wipes the token (this is by design — restore is the
///     recovery path, NOT persistence across reinstall).
public class DreamValleyAuthStorage: NSObject, FlutterPlugin {
  private static let methodChannelName = "com.vervetogether.dreamvalley/auth"
  private let service = "com.vervetogether.dreamvalley.auth"
  private let account = "session_token"

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = DreamValleyAuthStorage()
    let channel = FlutterMethodChannel(
      name: methodChannelName,
      binaryMessenger: registrar.messenger()
    )
    channel.setMethodCallHandler { call, result in
      instance.handle(call, result: result)
    }
  }

  private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "store":
      let args = call.arguments as? [String: Any] ?? [:]
      let token = (args["token"] as? String) ?? ""
      if token.isEmpty {
        NSLog("[DVAuth] store: empty token rejected")
        result(false)
        return
      }
      result(store(token: token))

    case "read":
      result(read())

    case "clear":
      result(clear())

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // MARK: - Keychain ops

  private func store(token: String) -> Bool {
    guard let data = token.data(using: .utf8) else {
      NSLog("[DVAuth] store: utf8 encode failed")
      return false
    }

    // Delete any existing entry first. SecItemUpdate has historically had
    // edge cases on accessibility-flag changes; delete + add is safer.
    let deleteQuery: [String: Any] = [
      kSecClass as String: kSecClassGenericPassword,
      kSecAttrService as String: service,
      kSecAttrAccount as String: account,
    ]
    SecItemDelete(deleteQuery as CFDictionary)

    let addQuery: [String: Any] = [
      kSecClass as String: kSecClassGenericPassword,
      kSecAttrService as String: service,
      kSecAttrAccount as String: account,
      kSecValueData as String: data,
      kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly,
    ]
    let addStatus = SecItemAdd(addQuery as CFDictionary, nil)
    if addStatus != errSecSuccess {
      NSLog("[DVAuth] store: SecItemAdd failed status=\(addStatus)")
      return false
    }

    // Write-back-verify: read immediately, confirm match. A silent Keychain
    // write that doesn't take effect is the exact invisible bug we guard
    // against — only return success if we can read back what we just wrote.
    let readBack = read()
    if readBack == token {
      NSLog("[DVAuth] store: success (write-back verified, token_len=\(token.count))")
      return true
    }
    NSLog("[DVAuth] store: write-back MISMATCH wrote_len=\(token.count) read_len=\(readBack?.count ?? -1)")
    return false
  }

  private func read() -> String? {
    let query: [String: Any] = [
      kSecClass as String: kSecClassGenericPassword,
      kSecAttrService as String: service,
      kSecAttrAccount as String: account,
      kSecReturnData as String: true,
      kSecMatchLimit as String: kSecMatchLimitOne,
    ]
    var result: AnyObject?
    let status = SecItemCopyMatching(query as CFDictionary, &result)
    if status == errSecItemNotFound {
      return nil
    }
    if status != errSecSuccess {
      NSLog("[DVAuth] read: SecItemCopyMatching failed status=\(status)")
      return nil
    }
    guard let data = result as? Data, let token = String(data: data, encoding: .utf8) else {
      NSLog("[DVAuth] read: data decode failed")
      return nil
    }
    return token
  }

  private func clear() -> Bool {
    let query: [String: Any] = [
      kSecClass as String: kSecClassGenericPassword,
      kSecAttrService as String: service,
      kSecAttrAccount as String: account,
    ]
    let status = SecItemDelete(query as CFDictionary)
    if status == errSecSuccess || status == errSecItemNotFound {
      NSLog("[DVAuth] clear: success (status=\(status))")
      return true
    }
    NSLog("[DVAuth] clear: SecItemDelete failed status=\(status)")
    return false
  }
}
